reconcile/openshift_resources_base.py

```diff
@@ -944,6 +959,209 @@ def run(
     return ri
 
 
+class CheckError(Exception):
+    pass
+
+
+class CheckNamespaceResources(Protocol):
+    def check(self) -> list[Exception]:
+        pass
+
+
+@dataclass
+class CheckClusterScopedResourceNames:
+    oc_map: OC_Map
+    ri: ResourceInventory
+    namespaces: list[Mapping[str, Any]]
+    thread_pool_size: int = 10
+
+    def check(self) -> list[Exception]:
+        errors: list[Exception] = []
+        for ns in self.namespaces:
+            cluster_name = ns["cluster"]["name"]
+            try:
+                oc = self.oc_map.get_cluster(cluster_name)
+            except OCLogMsg as ex:
+                if ex.log_level >= logging.ERROR:
+                    self.ri.register_error()
+                    errors.append(ex)
+                continue
+
+            ns_type_overrides = ob.get_namespace_type_overrides(ns)
+            managed_resource_types = ob.get_namespace_resource_types(
+                ns, ns_type_overrides
+            )
+            cluster_scoped_types = [
+                k for k in managed_resource_types if not oc.is_kind_namespaced(k)
+            ]
+
+            if len(cluster_scoped_types) > 0:
+                # Check that all non namespaced resources are explicitly set in the
+                # ManagedResourceNames attribute.
+                mrn = ob.get_namespace_resource_names(ns, ns_type_overrides)
+                for kind in cluster_scoped_types:
+                    declared_items = mrn.get(kind, [])
+                    desired_items = set(
+                        self.ri.get_desired_by_type(
+                            cluster_name, ns["name"], kind
+                        ).keys()
+                    )
+                    diff = desired_items.difference(declared_items)
+                    if len(diff) > 0:
+                        errors.append(
+                            CheckError(
+                                "Cluster scoped resources not defined in ManagedResourceNames. "
+                                f"cluster: {cluster_name}, namespace: {ns['name']}, "
+                                f"kind:{kind}, names:{diff}",
+                            )
+                        )
+
+        return errors
+
+
+@dataclass
+class CheckClusterScopedResourceDuplicates:
+    oc_map: OC_Map
+    all_namespaces: Optional[list[Mapping]] = None
+    thread_pool_size: int = 10
+
+    def check(self) -> list[Exception]:
+        errors: list[Exception] = []
+        clusters = set(self.oc_map.clusters() + self.oc_map.clusters(privileged=True))
+        cluster_cs_resources = get_cluster_scoped_resources(
+            self.oc_map, clusters, self.all_namespaces
+        )
+        results = threaded.run(
```

Comment:
_remap_cluster_result is  iterating over a list, without side effects, which benefit does threading provide here?

Reply:
The idea is to run the function in parallel for each cluster to do the process faster. Does that make sense?

Reply:
I was just wondering if it's really worth spawning threads for this given it should take only a few milliseconds.